### ISSUE TITLE
Request comment: webapp directory layout

### ISSUE THREAD

**Issue by erikkemperman** [jhipster/generator-jhipster#2869]
There has been some [discussion](https://github.com/jhipster/generator-jhipster/issues/2282) about sourcemaps, digressing into the size of the war, and wondering if, now that grunt was dropped, it might make sense to use `gulp-inject` instead of `wiredep` and `usemin`.

The more I think about it, the more I feel this might all be a lot more straightforward if the directory layout were a little different. Well, quite different to be honest...

I'm having fun playing around with an idea, but am acutely aware that it is a pretty sharp turn... So I am curious if anyone thinks that something like this would even be worth pursuing.

Let me try to explain what I'm thinking:

Currently, the production build in `dist` is a child of the source, `webapp`, which is essentially the  development build. Not quite, because there is some preprocessing (wiredep, ngconstants, sass). 

This has a few disadvantages, it seems to me.

- Sourcemaps of the non-embedded persuasion (which we would mostly prefer since  the development build is packaged along anyway) become unwieldy, they have  to somehow refer to files not properly in their webroot (cf., the servlet  filter which prepends `dist` to request paths in production).
- Also contained in `webapp` is the entire repository of bower components, most which shouldn't end up in the war file.
- Wiredep and ngconstants changes source files, more to the point versioned files. This is mildly annoying because it is easy to pollute commits with local constants or an index.html with injected bower refs.
- Likewise, sass overwrites css files in `webapp/content/css`, which are then versioned but IMHO, being just build artifacts, should not be.

All of this would be "remedied" by changing things around and having the result of production builds be the _parent_ of the development build, and if neither of them overlaps with versioned or bowered material.

So instead of prepending `dist` to the path in production, the servlet filter would prepend `dev` in development.

And the webbuild would put constants, injected html, sass-compiled stylesheets outside of versioned territory.

We might have to selectively copy some miscellaneous assets, but only to the extent that we can't make our build reliably identify what is needed automatically.

And, oh yeah, sourcemap references become trivial.

Sorry about the wall of text, but looking forward to your thoughts!

**Comment by deepu105** [comment]
That seems interesting, can you layout the directory structure you have in
mind?  If it improves things then definitely we can consider as we are
breaking things with this build anyway

Say :

src
    |__webapp
        |__so on

**Comment by erikkemperman** [comment]
> can you layout the directory structure you have in mind? 

Good idea, should have done that in the first place. I am thinking something like this:

```sh
src/main
 ├─ websrc                   # <-- versioned
 │   ├─ index.html
 │   ├─ bower_components
 │   ├─ content
 │   │   ├─ app      
 │   │   └─ css
 │   ├─ scss
 │   └─ gulpfile.js
 ├─ webapp                   # <-- deployed, .gitignore'd
 │   ├─ dev
 │   │   ├─ index.html       # <-- injected
 │   │   ├─ bower_components # <-- subset
 │   │   └─ content
 │   │       ├─ app          # <-- ngconstants, and copies of main scripts
 │   │       └─ css          # <-- sass-compiled, and copies of main styles
 │   ├─ index.html           # <-- minified
 │   ├─ content              # <-- concatentated, minified, sourcemapped
 │   │   ├─ app
 │   │   └─ css
 ... 
```

The build would basically copy/preprocess stuff from `websrc` to `webapp/dev`, and on top of that minify, concatenate and so on into `webapp`. Obviously these new names are to be determined, but you get the idea I hope.

**Comment by deepu105** [comment]
Thats Interesting. Lets see what other have to say
@jhipster/developers

**Comment by andidev** [comment]
I agree that the current solution where Wiredep, ngconstants and sass overwrites the source files. I think this sounds like a good improvement! How is all the other generators doing it (e.g. angular fullstack), anyone with insight?

**Comment by deepu105** [comment]
@erikkemperman the only problem is its gonna be a hell lot of refactor on the same scale as what I did for the app folder restructure and thats a lot of work

**Comment by erikkemperman** [comment]
Well from playing around with this it actually seems worse than it is. Mostly the gulpfile will change, but then that's what I started looking at (use gulp-inject instead of wiredep/usemin). Much of the rest is search/replace.

I'm certainly encouraged to continue experimenting, sounds so far like you guys at least don't think the idea is entirely mad :-)

I would be interested to learn how other generators lay out their builds.

**Comment by gmarziou** [comment]
It means also that you would have to change `StaticResourcesProductionFilter`.

If we were strictly following maven conventions, production files should not be generated under `src/main/webapp` but under `target/xxx` but it is probably harder.

**Comment by erikkemperman** [comment]
True, but I figured out how to do that already. Not too hard. And actually it seems to me that it would become slightly better -- rather than do a little extra work for each request in production, as it does now, this filter would do instead a little extra work for each request in developent. I guess the classname might have to change to `StaticResourcesDevelopmentFilter` :-)

**Comment by jdubois** [comment]
Agreed, it also looks more logical that we have the "normal" folder for production, and something specific for development.
Anyway, the best comment I've seen here is how do the other generators handle this? We are certainly not the first ones with this issue, and people from Angular Fullstack (typically) are much better than us for this.

**Comment by erikkemperman** [comment]
I'm not too familiar with Angular Fullstack, but it looks like it uses grunt. It is my impression that the wiredep/usemin setup is the traditional approach there, but it seems to be a slightly awkward fit for a streaming build system like gulp.

Or more precisely, I guess, is that the corresponding gulp plugins, especially gulp-usemin, suffers from slightly lax maintainance (no offence to its authors!) and some quirks in edge cases (like the one, I suspect, injecting sourcemaps into the html as `<link>` stylesheets). 

About the directory structure, I guess JHipster is different from something like Fullstack in that
- it inherits the `src/main/webapp` convention from Java/Maven practice;
- it inherits the convention from webdev practice (mostly html5-boilerplate, I suppose) of having a `dist` directory as a child of source;
- together this means we are putting build artifacts in what is, relative to the overall project root, a source directory;

Maybe it makes sense, like @gmarziou pointed out, to make the web-build operate on `target/xxx` directly. That way we can keep the path `src/main/webapp` as is (no need for the awkward `src/main/websrc` I suggested). Would that then work for Gradle projects as well?

**Comment by deepu105** [comment]
Gradle doesnt follow the target directory, but im not sure it its possible
to tweak that @atomfrede your view on this plz

**Comment by deepu105** [comment]
I looked at few generators and none of them offers complicated production config and dev config like we do so we may have to come up with something here

@erikkemperman if we do this restructure what are the benefits that we are looking at other than having sourcemaps (with mangling or without mangling?)
btw @erikkemperman they have gulp as well [here](https://github.com/angular-fullstack/generator-angular-fullstack/blob/master/app/templates/gulpfile.babel(gulp).js)

btw @jdubois I think Jhipster now does much more than what Angular fullstack does and does some of it way better too IMO :wink:

**Comment by jdubois** [comment]
Yes I'm still surprised by how far we go!!

**Comment by erikkemperman** [comment]
@deepu105
Well, sourcemaps with mangling is possible with current structure -- as in that branch I linked earlier. That just has a quirk I coulant track down quickly but suspect it's in gulp-usemin.

We could try to change to gulp-inject without restructuring layout.

But what I proposed seems more logical to me, and would certainly make source map configuration much easier.

The main benefit to my mind would be no build products in source (versioned) directory, on the one hand, and no unneeded source material (most of bower_components) in deploy/package directory, on the other.

**Comment by erikkemperman** [comment]
@deepu105 Ah now I notice fullstack gulp support, thanks! Didn't read far enough down their readme, which unconditionally says to install grunt. I will take a look at their structure when I get a chance.

**Comment by erikkemperman** [comment]
Actually the fullstack gulpfile is pretty instructive (once you get past that it's ES6)!

It uses `lazypipe` for re-usable pipelines, which is probably what `gulp-usemin` should have been using so we wouldn't have to do otherwise useless `bind`ing of plugins.

Also it uses `gulp-inject` instead of `gulp-usemin` but then it did keep `gulp-wiredep` for some reason (which I thought might also be replaced by `gulp-inject`, combined with `main-bower-files`). Perhaps it does this better in some way I don't yet understand, or perhaps it is a remnant of a port from  grunt to gulp.

It does appear to keep the traditional approach of nesting `dist` as a child of the webroot, although that's just from reading the gulpfile, I haven't yet had an opportunity to actually run the generator. But tentatively I would still guess that JHipster, with its Java (servlet container) backend, will probably end up having different criteria for what makes a good directory layout...

**Comment by andidev** [comment]
I'm not sure we can replace gulp wiredep with gulp inject and the main files since it also needs to determine the order of the injected dependencies and that is not part of gulp inject unless I'm wrong. 

I think gulp wiredep is used to inject Bower libs while gulp inject is used for injecting application javascript files. Please correct me If I'm wrong.

**Comment by atomfrede** [comment]
@deepu105 I don't see any reason why using the build folder (or just use target folder as build output folder) shouldn't work with gradle.

**Comment by erikkemperman** [comment]
@andidev I thought `gulp-inject` is rather generic, which ought to allow it do dependency-injection from bower as well. It has examples of usage with bower pretty early in its docs. Of course it only works so long as it injects in the right order -- maybe you're worried `main-bower-files` doesn't follow sub-dependencies? I dn't know, have just been playing around with a trivial case thusfar.

@atomfrede That's great, it would mean we could just have gulp write to `/target` straight away. Unless... It kind of takes over a part of what the maven (and I guess gradle) build would normally do, namely moving stuff from `src/main/webapp` to `target`. That might be an argument in favour of my original tree outline above, i.e. that the maven/gradle package/deploy phase doesn't deviate from convention?

**Comment by deepu105** [comment]
Im ok with this as long as following conditions are met and if you are upto
it, trust me it would be a lot of work even if you are just doing search
and replace, you will get un expected errors and will be doing a lot of
corrections.

1. This shouldn't slow down our builds. maven, gradle and gulp
2. App should work without issue when run directly or as war/jar in both
dev and prod profile
3. Behaviour should be same for maven and gradle
4. Browsersync should work

Last but not least we need to update documentation as well.

So if other hipsters agree you can proceed

**Comment by erikkemperman** [comment]
I would suggest untangling a possible directory layout change from swapping out gulp plugins.

Obviously, I will try my best to get it as solid as I know how -- but have to admit upfront that there are certain areas I don't know much about. In particular, things like Heroku/Cloudfoundry, also Gradle. So I would definitely need thorough review on those things, from people more knowledgable than me...

To the extent that the refactor would be a lot of work... Yeah, I noticed there's a lot of literal instances of `src/main/webapp` :-) First thing I would do is use shared constants for that. Once that is done it really should be pretty straightforward, but I know... Famous last words.

**Comment by deepu105** [comment]
It is already a shared constant, may be we should use the constant in the
templates as well instead of literal string.

The paths are used in
1. Main generator index.js
2. Sub generators index.js
3. All the angular route files
4. In some controllers to open popups
5. Tests
6. Gulp file

I might have missed some.

Another problem is such a huge change will not show all diffs in git and is
tough to review so if others agree and you proceed ill suggest you to do
many small PR dealing one step at a time so that we can review properly

**Comment by erikkemperman** [comment]
I noticed the shared constant, but it isn't used consistently -- unless I am missing something here? Grep for the literal `'src/main/webapp'` turns up rather a lot. I would make them available in template strings as well.

**Comment by erikkemperman** [comment]
I'm not sure how to do something like this one small step at a time -- it strikes me as a kind of all-or-nothing approach. None of the intermediate steps would be functional.

**Comment by atomfrede** [comment]
@deepu105 At work we've used the angular-gulp-generator and put the source files into a sperate drirectory (``src/web``). Bower, node etc. are put into ``target`` (when using maven) and the maven build just takes the files generated by gulp (and some small configs from ``src/main/webapp`` like filter configs) and puts them into the generated war. Same should be possible for gradle I guess (using ``build`` folder)

**Comment by andidev** [comment]
@erikkemperman yes you are right main bower files should return the dependencies in order of course! Did not know about that project.

**Comment by deepu105** [comment]
Ill open a new issue for gulp inject @erikkemperman may be you can do gulp inject as a separate PR and later if others agree you can do the restructuring what say?

**Comment by erikkemperman** [comment]
Yeah, that's what I meant by untangling earlier. However it seems more sensible to me to do this restructuring first -- unless there are objections forthcoming, of course.

**Comment by erikkemperman** [comment]
Update -- so, I went ahead and tried to get most of the tedious things out of the way. This [branch](https://github.com/erikkemperman/generator-jhipster/tree/refactor-webapp) passes `npm test` on the generator and `gulp test` on the generated app, which seems to be working nicely for both dev and prod profiles.

I've mostly tested with sass and i18n enabled, only with Maven... Clearly there's a lot to polish and there will certainly be some gaps and bits I just didn't notice. Travis fails, obviously.

But since I won't have much time for two, maybe three days, I figured this might be a good time to share what I have...

If anyone feels this is a really bad idea or I'm doing it wrong, or wants to pitch in, please get in touch.

**Comment by gmarziou** [comment]
I just tried briefly, gulp test works, mvn spring-boot:run also.
But when I try to run both gulp and mvn, the angular app does not work but I don't see errors in the console the first time, if I reload the page I get 404 errors for following i18n files:

- /dev/bower_components/angular-i18n/angular-locale_en.js 
- /dev/i18n/en/global.json
- /dev/i18n/en/home.json

Here is my config:

~~~json
{
  "generator-jhipster": {
    "jhipsterVersion": "2.27.0",
    "baseName": "newLayout",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "sql",
    "devDatabaseType": "h2Memory",
    "prodDatabaseType": "postgresql",
    "searchEngine": "no",
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "rememberMeKey": "912bbb0681b86009d4959f9bee7041793a500546",
    "useSass": true,
    "enableTranslation": true,
    "applicationType": "monolith",
    "testFrameworks": [
      "gatling",
      "cucumber",
      "protractor"
    ]
  }
}
~~~

**Comment by erikkemperman** [comment]
@gmarziou Thanks for taking a look! Interesting, I had i18n working; The files you mention are present in my webapp/dev folder. So it looks like `gulp prepare` did not copy these for you? Do they exist in websrc?

**Comment by deepu105** [comment]
closed via #2945
